### PR TITLE
Change servo parameters to use ros2 interfaces for parameters

### PR DESCRIPTION
**Pull request by stephanie-eng**
### Description
Update servoParameters to use a new declare and get function that uses NodeParametersInterface to declare and get parameters. See #790

This PR  does the following:
- Sets default values for servo parameters based off of panda_simulated_config.yaml
- Replaces the getOrDeclareParam function with individual get and set functions
  - These 
- Adds parameter_descriptor_builder, which is used to create the messages used in setting up the descriptions for parameters

TODO:

- [x] When running the joystick servo example, declare will throw an error saying that the parameters are already declared. Not sure why this is happening yet, still working on this.
- [x] Build error with the set parameters callback

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (stephanie-eng) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:442]: When running the joystick servo example, this will throw an exception that suggests the paramaters have already been declared before the declare function gets to do it. Catching this right now just as a test. Will try to fix later.

File: moveit_ros/moveit_servo/src/parameter_descriptor_builder.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/parameter_descriptor_builder.cpp:63]: You should be able to uncomment this.  Also, please look into the ParameterDescriptor to make sure there are not any new features that have been added since I did this.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:433]: Use make_shared here.  Reference: https://stackoverflow.com/questions/20895648/difference-in-make-shared-and-normal-shared-ptr-in-c

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:43]: This should be <moveit_servo/parameter_descriptor_builder.hpp>

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:85]: What does this mean?

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:270]: This comment seems out of place.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:268]: Instead of using an output parameter here, this should just return a new ServoParameters object.  Something like this should work:

```suggestion
ServoParameters ServoParameters::get(const std::string& ns,
                          const rclcpp::node_interfaces::NodeParametersInterface::SharedPtr& node_parameters)
```

Then since this function is a member of the class it should be able to use the private constructor.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:83]: Similar to the get function, this shouldn't need parameters shared ptr parameter.  Also, as declaring does not set any values (it only describes the parameter interface to the ros system) it shouldn't set values in a parameters object.  Internally you should just create a temporary one for the default values.  Something like this should work:

```suggestion
void ServoParameters::declare(const std::string& ns,
                              const rclcpp::node_interfaces::NodeParametersInterface::SharedPtr& node_parameters)
{
  auto parameters = ServoParameters{ns};  // for populating default values
```

Lastly, you should be able to get rid of the logger out of the constructor and just put one in this file's scope similar to what is done in other source files in moveit_servo.

**Review by tylerjw — COMMENTED** [review]
I left a few comments, let me know if you have questions.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (stephanie-eng) [reply moveit_ros/moveit_servo/src/servo_parameters.cpp:85]: See https://github.com/ros-planning/moveit2/blob/b426a1c6f957522f8261f3500ee11f82199cdef6/moveit_ros/moveit_servo/src/servo_parameters.cpp#L91

I believe is saying that the parameters below are properties of incoming commands (ex. command_in_topic)

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (stephanie-eng) [reply moveit_ros/moveit_servo/src/servo_parameters.cpp:272]: From here:
https://github.com/ros-planning/moveit2/blob/b426a1c6f957522f8261f3500ee11f82199cdef6/moveit_ros/moveit_servo/src/servo_parameters.cpp#L91-L99

I believe is saying that the parameters below are properties of incoming commands (ex. command_in_topic)

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (stephanie-eng) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:453]: TBD: Uncomment when this stops causing a build error

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (stephanie-eng) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:435]: TBD: Probably can remove this now that we don't need logger in the constructor. Removing this will mean we have to edit function calls in other files as well.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (AndyZe) [reply moveit_ros/moveit_servo/src/servo_parameters.cpp]: ^yep, exactly

**Comment by AndyZe** [comment]
In my first time testing this, it fails due to an undeclared parameter. It's nice that it failed and printed a useful error message but the error message could be formatted better:

```
[servo_node_main-3] terminate called after throwing an instance of 'rclcpp::exceptions::ParameterNotDeclaredException'
[servo_node_main-3]   what():  moveit_servocommand_out_topic
```

Can you please update this with a bit of a description of `moveit_servocommand_out_topic`? Otherwise it is easy to miss.

```
[servo_node_main-3] terminate called after throwing an instance of 'rclcpp::exceptions::ParameterNotDeclaredException'
[servo_node_main-3]   Parameter name:  moveit_servocommand_out_topic
```

File: moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.h
Comment (AndyZe) [inline moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.h:46]: Nitpick: please alphabetize. I think the order should be: system includes, then MoveIt includes.

```
#include <mutex>
#include <thread>
#include <vector>

#include <moveit/planning_scene_monitor/planning_scene_monitor.h>
#include <rclcpp/rclcpp.hpp>
```

**Review by AndyZe — COMMENTED** [review]
@tylerjw and @stephanie-eng can you please add to the PR description -- **why** should we use parameter_descriptor_builder? I guess it has to do with setting parameters dynamically.

Without a better PR description, it kind of looks like 500 lines of code have been added without much change in functionality.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (AndyZe) [reply moveit_ros/moveit_servo/src/servo_parameters.cpp:278]: ^Stephanie's right.  :+1:

**Comment by stephanie-eng** [comment]
> In my first time testing this, it fails due to an undeclared parameter. It's nice that it failed and printed a useful error message but the error message could be formatted better:
> 
> ```
> [servo_node_main-3] terminate called after throwing an instance of 'rclcpp::exceptions::ParameterNotDeclaredException'
> [servo_node_main-3]   what():  moveit_servocommand_out_topic
> ```
> 
> Can you please update this with a bit of a description of `moveit_servocommand_out_topic`? Otherwise it is easy to miss.
> 
> ```
> [servo_node_main-3] terminate called after throwing an instance of 'rclcpp::exceptions::ParameterNotDeclaredException'
> [servo_node_main-3]   Parameter name:  moveit_servocommand_out_topic
> ```
> 
> Or incorporate it all in one line. That would be even better.
> 
> `[servo_node_main-3] terminate called after throwing an instance of 'rclcpp::exceptions::ParameterNotDeclaredException'. Parameter name: moveit_servocommand_out_topic`

This was due to me having a typo with some of the parameters. The parameter name causing this issue is hard-coded, so now that it's been corrected, is it still worth to have a specific error message?

**Comment by tylerjw** [comment]
@AndyZe I'll respond to the parameter descriptor builder questions.  Long term I hope to put something like this in a much more general place than in servo.  It has to do with adding descriptions and types to the parameter declarations.  

ROS2 has this object (ParameterDescriptor) that it can take in the declare_parameter functions in order to provide a description and other details about the parameter that the ros2 system can use to help the user when interacting with the node.  Unfortunately, because this object is a ROS2 message the code is much longer and uglier if you don't have the ParameterDescriptorBuilder here.  An example of it in ROS2 is here: https://github.com/ros2/rclcpp/blob/1037822a63330495dcf0de5e8f20544375a5f116/rclcpp/include/rclcpp/detail/qos_parameters.hpp#L163  The ParameterDescriptorBuilder makes this much nicer when you have to create a bunch of these (because you have a bunch of parameters).

One large advantage of doing this is a user will be able to run `ros2 param dump <servo_node_name>` and it will output a yaml file of the parameters.  Another thing you can run is `ros2 param list <servo_node_name>` and it will output a list of the parameters and their descriptions.

This is also needed to support dynamic reconfiguring in the future.  As to the features of ParameterDescriptor, additional constraints as I understand it is a string where you can describe things like how the values set in that parameter affect other parameters.  You might put something like "must be a valid ros topic name" or "only considered if <other parameter> is true" or something like that.

**Comment by AndyZe** [comment]
> This was due to me having a typo with some of the parameters. The parameter name causing this issue is hard-coded, so now that it's been corrected, is it still worth to have a specific error message?

No worries, I know this is still a draft. What I'm wondering is, can you somehow print the missing parameter name in the same line as the rest of the error message? In a general way, no matter what parameter it is. If not, it's not a huge deal. Like this:

`[servo_node_main-3] terminate called after throwing an instance of 'rclcpp::exceptions::ParameterNotDeclaredException'. Parameter name: whatever_my_parameter_name_is`

File: moveit_ros/moveit_servo/include/moveit_servo/parameter_descriptor_builder.hpp
Comment (henningkayser) [inline moveit_ros/moveit_servo/include/moveit_servo/parameter_descriptor_builder.hpp:97]: Should this function be camel case like the others?

**Review by henningkayser — COMMENTED** [review]
First off, I think this is a really good improvement as this approach provides much more consistency. I'd really appreciate if we could remove `automatically_declare_parameters_from_overrides` in all nodes. I don't have much to add to this PR as this is a step into the right direction.

However, I have plenty of feedback and questions that address the new helper API and this use case is a good opportunity to continue the design discussions we started half a year ago (again, this is not meant to block this PR, we should even get this merged imo):

* There still is way too much boilerplate code and unnecessary duplication, inviting for user errors and frustrating debugging sessions (overstated, but still this this should be a lot simpler and shorter)
* The descriptor constraints are somewhat useless if they don't provide the implementation for the `validate` function, `additional_constraints` are even more useless in this context
* The application code is responsible for handling parameter callbacks while the config itself is not being updated
* Intermixed runtime dependencies on rclcpp, this doesn't seem to go well with mock tests
* No support for parameter namespaces, let alone group parameter constraints which are de facto implemented in the validate function.

I'd really like to move towards an API that meets all of these requirements. While I agree with @tylerjw that we shouldn't abstract completely from rclcpp, it's unlikely that rclcpp will provide these any time soon, unless we implement something that works for us first and try to bring it upstream. The goal should really be something that is just as easy to use as `rosparam_shortcuts` while allowing us to work on the actually challenging problems like:
* Support for config profiles and generic parameter types (=parameter groups)
* Synchronization of parameter groups (e.g. transforms, profiles, robot model etc)
* Validating, versioning and generating of (yaml) config files

With that, I think it makes sense to continue the discussion in #790, https://github.com/PickNikRobotics/parameter_tools, or even https://github.com/PickNikRobotics/ros2_params. I'll put in more time to rework some of the interfaces and make suggestions on how to simplify things without abstracting too much.

**Comment by henningkayser** [comment]
> > ```
> > * No support for parameter namespaces
> > ```
> 
> Does this work with namespaces? That's an absolute must. We have a dual-arm robot project starting up right now.

It does, but you have to stitch them together like ```ns + ".parameter"```

File: moveit_ros/moveit_servo/include/moveit_servo/parameter_descriptor_builder.hpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/include/moveit_servo/parameter_descriptor_builder.hpp:88]: Please remove this comment.

File: moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.h
Comment (tylerjw) [inline moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.h:59]: This isn't needed anymore, so you should remove it.

File: moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.h
Comment (tylerjw) [inline moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.h:109]: Please don't put the pose tracking parameters in here.  Please create a separate pose tracking parameters object that uses the same pattern and declares/gets these.

File: moveit_ros/moveit_servo/src/parameter_descriptor_builder.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/parameter_descriptor_builder.cpp:30]: @AndyZe here is the description of this.  I will think of a nicer place to put this than in servo.  I think having builders for some commonly used message types in moveit would be fine and so I can create a package for these.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:89]:
```suggestion
  auto parameters = ServoParameters{};
```

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:94]: ParameterDescriptorBuilder doesn't need to be in a separate `parameter_set` namespace.  It can just be in the moveit_servo namespace for now.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:95]: You can make all of this easier to read with a few using statements at the top of this function for the types.  For example:

```
using rclcpp::ParameterValue;
using rclcpp::ParameterType::PARAMETER_BOOL;
using rclcpp::ParameterType::PARAMETER_STRING;
```

**Review by tylerjw — COMMENTED** [review]
A couple of small style things, but otherwise this looks really good.  I'll test this later today to verify it doesn't change the behavior.

Have you tried running the tutorial after these changes?

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:98]: I'm sorry to be nit-picky about this but we should use brace initialization by default.

```suggestion
      ns + ".use_gazebo", ParameterValue{parameters.use_gazebo},
```

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:99]:
```suggestion
      ParameterDescriptorBuilder{}
```

**Comment by stephanie-eng** [comment]
> Have you tried running the tutorial after these changes?
I've been running the joystick and pose tracking examples from the launch folder as a sanity check. I don't have a joystick handy, so I can't move the arm for the teleop demo, but otherwise seems to work.

File: moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.h
Comment (tylerjw) [inline moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.h:79]: Do you mind using uniform initialization for these too?

File: moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.h
Comment (tylerjw) [inline moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.h:132]:
```suggestion
      const std::lock_guard<std::mutex> guard{callback_handler_->mutex_};
```

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:253]:
```suggestion
  auto parameters = ServoParameters{};
```

**Comment by AndyZe** [comment]
>  I don't have a joystick handy

@stephanie-eng the tutorial has a way to test with a keyboard. (https://github.com/ros-planning/moveit2_tutorials/blob/main/doc/realtime_servo/realtime_servo_tutorial.rst)

ros2 launch moveit_servo servo_example.launch.py
ros2 service call /servo_node/start_servo std_srvs/srv/Trigger {}
ros2 run moveit2_tutorials servo_keyboard_input

**Comment by stephanie-eng** [comment]
Thanks, Andy.
I've incorporated Tyler's suggestions and was able to run the demo without problem.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:107]: You don't need the rclcpp namespace for ParameterValue here because of the using statement.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:110]: Here too

**Review by tylerjw — COMMENTED** [review]
Only a couple of tiny things I found.  Otherwise this looks good.  I tested this locally and it doesn't break anything as far as I can tell.

**Comment by stephanie-eng** [comment]
Just as an FYI: because this changes the function signature of makeServoParameters, a minor update to the tutorial is required for it to build. I've created a PR to make the tutorials build again for after this PR is merged.

**Comment by tylerjw** [comment]
> Just as an FYI: because this changes the function signature of makeServoParameters, a minor update to the tutorial is required for it to build. I've created a PR to make the tutorials build again for after this PR is merged.

Thank you for doing that.  :)

**Comment by AndyZe** [comment]
moveit_ros/moveit_servo/include/moveit_servo/parameter_descriptor_builder.hpp could still use a description at the top of the file. Something like:

```
/* Author: Stephanie Eng
   Description: blah blah blah
 */
```

**Review by AndyZe — APPROVED** [review]

**Pull request merged**
